protocols/gossipsub/src/handler.rs

```diff
@@ -163,6 +167,10 @@ impl GossipsubHandler {
             peer_kind_sent: false,
             protocol_unsupported: false,
             upgrade_errors: VecDeque::new(),
+            // mxinden: In the case where another NetworkBehaviour requests a connection to a remote
```

Comment:
I don't think so. Maybe we could make this more explicit however. 

When a new connection is established, gossipsub will try to send any subscriptions to the peer. To do this it forces a negotiation attempt. If the negotiation fails then the keep_alive gets adjusted. 

There is however the edge case that locally we are not subscribed to any topics. In this case we don't send/receive any messages on gossipsub messages but the handler maintains the keep-alive. As soon we subscribe to anything or get a message from a peer on gossipsub we have to start the negotiation. 

Perhaps this should be set to a timeout limit to handle this case?

Reply:
> Perhaps this should be set to a timeout limit to handle this case?

Yes, from what I can tell a timeout would take care of most edge-cases. Keep-alive handling has cost us many debugging hours in the past (e.g. see https://github.com/libp2p/rust-libp2p/pull/1698). I tried to simplify the concept with https://github.com/libp2p/rust-libp2p/pull/1717, but without success thus far. Happy for input in case you have ideas.